element-connector/src/main/java/org/eclipse/californium/elements/RawData.java

```diff
@@ -225,13 +227,15 @@ public static RawData inbound(byte[] data, InetSocketAddress address, Principal
 	 * 
 	 * @param data the data to send.
 	 * @param address the IP address and port the data is to be sent to.
+	 * @param correlationContext correlation context for sending data (may be <code>null</code>).
 	 * @param callback the handler to call when this message has been sent (may be <code>null</code>).
 	 * @param useMulticast indicates whether the data should be sent using a multicast message.
 	 * @return the raw data object containing the outbound message.
 	 * @throws NullPointerException if data or address is <code>null</code>.
 	 */
-	public static RawData outbound(byte[] data, InetSocketAddress address, MessageCallback callback, boolean useMulticast) {
+	public static RawData outbound(byte[] data, InetSocketAddress address, CorrelationContext correlationContext, MessageCallback callback, boolean useMulticast) {
```

Comment:
Sticking to your example, I guess the PSK identity would come from a `CorrelationContext` established when the client initially authenticated and issued its first request. The server then sends a response to the client using the `CorrelationContext` from the ongoing `Exchange` the response is part of, right? I think this case could be handled by simply providing the `CorrelationContext` instance from the `Exchange` to `RawData.outbound()`, it should contain everything that's needed to make sure that the outbound response is indeed matching the request.
However, when the server later wants to send a request to the client itself, it needs to make sure that the request is sent to the client over the same DTLS session that was established when the client sent its initial request. Doesn't the server need to somehow *cache* or *associate* the then established `CorrelationContext` with the client identifier (e.g. the client's IP:port or whatever)?
Is this the responsibility of the application layer? How does an application provide the information to the stack which is required for matching the `CorrelationContext` to a client security association?